src/DDMinimizer.cpp

```diff
@@ -0,0 +1,426 @@
+            itxH->second = instructionIndex;
+            found = true;
+          }
+        }
+      }
+    }
+    instructionIndex++;
+  }
+  if (!found) {
+
+    return qc.initialLayout;
+  }
+
+  // iterate over all the maps and find the max index of the found x-c pairs
+  cXIndex = DDMinimizer::findMaxIndex(cXMap);
+  xCIndex = DDMinimizer::findMaxIndex(xCMap);
+  for (size_t i = 0; i < bits - 1; i++) {
+    cLIndex[i] = DDMinimizer::findMaxIndex(cLMap[i]);
+    cHIndex[i] = DDMinimizer::findMaxIndex(cHMap[i]);
+    xLIndex[i] = DDMinimizer::findMaxIndex(xLMap[i]);
+    xHIndex[i] = DDMinimizer::findMaxIndex(xHMap[i]);
+  }
+
+  // create the permutation based on the order of max index in the complete maps
+  std::vector<Qubit> layout(bits);
+  std::iota(layout.begin(), layout.end(), 0);
+
+  const int prioCh = DDMinimizer::getLadderPosition(cHIndex, xCIndex);
+  const int prioXl = DDMinimizer::getLadderPosition(xLIndex, xCIndex);
+  const std::size_t stairsCh = DDMinimizer::getStairCount(cHIndex);
+  const std::size_t stairsXl = DDMinimizer::getStairCount(xLIndex);
+  const int prioCl = DDMinimizer::getLadderPosition(cLIndex, cXIndex);
+  const int prioXh = DDMinimizer::getLadderPosition(xHIndex, cXIndex);
+  const std::size_t stairsCl = DDMinimizer::getStairCount(cLIndex);
+  const std::size_t stairsXh = DDMinimizer::getStairCount(xHIndex);
+
+  // complete case checkins and adjust the layout
+  if ((cXIndex != -1) && (xCIndex == -1 || cXIndex < xCIndex)) {
+
+    if (prioCh == 0 && stairsCh > 0) {
+      layout = DDMinimizer::reverseLayout(layout);
+      layout = DDMinimizer::rotateLeft(layout, stairsCh);
+    } else if (prioXl == 0 && stairsXl > 0) {
+      layout = DDMinimizer::reverseLayout(layout);
+      layout = DDMinimizer::rotateRight(layout, stairsXl);
+    } else if (prioCh > 0 || prioXl > 0 || (prioCh == 0 && prioXl == 0)) {
+      layout = DDMinimizer::reverseLayout(layout);
+    }
+  } else if ((xCIndex != -1) && (cXIndex == -1 || cXIndex > xCIndex)) {
+
+    if ((prioCl == 0 && DDMinimizer::isFullLadder(cLIndex)) ||
+        (prioXh == 0 && DDMinimizer::isFullLadder(xHIndex))) {
+      layout = DDMinimizer::reverseLayout(layout);
+    }
+
+    else if (prioCl == 0 && stairsCl > 0) {
+      layout = DDMinimizer::rotateRight(layout, stairsCl);
+    } else if (prioXh == 0 && stairsXh > 0) {
+      layout = DDMinimizer::rotateLeft(layout, stairsXh);
+    }
+  } else if ((DDMinimizer::isFullLadder(xHIndex) ||
+              DDMinimizer::isFullLadder(cLIndex))) {
+    layout = DDMinimizer::reverseLayout(layout);
+  } else {
+    // in case no full pattern was identified, call fallback function to
+    // determine ordering based on singular controlled operations
+    return DDMinimizer::createControlBasedPermutation(qc);
+  }
+
+  // transform layout into permutation
+  qc::Permutation perm;
+  std::vector<Qubit> layer(bits);
+  for (qc::Qubit i = 0; i < bits; i++) {
+    layer[i] = i;
+  }
```

Comment:
I do not understand the use of `layer`. If I am not mistaken, `layer` serves as an identity mapping.

If you have trouble with converting `std::size_t` to `qc::Qubit` you can use cast functions, e.g., `static_cast<qc::Qubit>(...)`.